feat: Add Flutter Core Support for GDPR and CCPA

BrandonStalnaker opened the pull request:
# Summary
Adds dart support for GDPR and CCPA and implements them in the iOS layer.

Tested with livestream after updating sample app

TP: #68249

rmi22186 reviewed (changes requested):
When pulling the code and testing, it seems like the removeGDPR and removeCCPA functions aren't working.  See video capture here - https://share.getcloudapp.com/DOuByB1J

rmi22186 commented on lib/src/user.dart:
```diff
@@ -92,4 +94,68 @@ class User {
   }
+
+  /// Returns all GDPR Consent states with their purpose.
+  Future<Map> getGDPRConsentState() async {
```
I think we should import GDPRConsent and set this to return:
```suggestion
  Future<GDPRConsent> getGDPRConsentState() async {
```

rmi22186 commented on lib/src/user.dart:
```diff
@@ -92,4 +94,68 @@ class User {
+  }
+
+  /// Returns the CCPA Consent state.
+  Future<Map> getCCPAConsentState() async {
```
similar to above comment but instead of a map it should be just CCPAConsent since only 1 can exist 
```suggestion
  Future<CCPAConsent> getCCPAConsentState() async {
```

BrandonStalnaker commented:
Updated that method to return the objects and corrected the sample app. Meanwhile as we spoke about in standup the remove issue seems to be from core. Let me know if that occurs when your implementing in web.

rmi22186 reviewed (approved):
lgtm.  i'd still like a second look at the API from @willpassidomo and a review of the swift code from @peterjenkins .  This does function as expected when pulling down the new code.

peterjenkins reviewed (commented):
Swift code looks good from what I can see

einsteinx2 reviewed (dismissed):
Reviewed the Swift code, cleaned up some formatting and fixed a few minor warnings, then tested the example iOS app. LGTM

willpassidomo commented:
@rmi22186 ok, so I really like just following the SDK's API, but I think in this case there's no reason to have a separate `GDPRConsent` and `CCPAConsent` objects. They're identical and there's no reason why `setCCPAConsentState()` and `setGDPRConsentState()` can't just both take in `Consent` (or `ConsentInstance`?..idk, the plain one makes more sense)

on Android we kept `GDPRConsent` when CCPA was implemented just because it was already out there, but I think we pretty clearly bungled the initial design...I mean, who could have thought there'd be more than one piece of consumer-privacy legislation (lol)

rmi22186 commented on ios/Classes/SwiftMparticleFlutterSdkPlugin.swift:
```diff
@@ -12,7 +12,7 @@ public class SwiftMparticleFlutterSdkPlugin: NSObject, FlutterPlugin {
     switch call.method {
     case "isInitialized":
-        result(MParticle.sharedInstance() != nil)
+        result(true)
```
I think we want to revert this change. We want to know if there's a shared instance or not, and not always return true.

BrandonStalnaker replied:
The reason @einsteinx2 changed this is because of the singleton pattern we will always have an instance and it creates a warning to have the code from before.

einsteinx2 replied:
Yes exactly, the instance is always there so for iOS it's actually always initialized. That previous line caused a warning since it's always evaluated to true, so I just made it return true.

willpassidomo replied:
So, I think this was probably bad logic to begin with, but `isInitialized` _should_ be returning a boolean based on whether the SDK has been started (aka has `start()` been called and are we good to log events and such)

BrandonStalnaker replied:
OK got it. I think we were confirming whether or not the instance was available. Updated to check if initialized.

rmi22186 commented on example/lib/main.dart:
```diff
@@ -382,6 +383,84 @@ class _MyAppState extends State<MyApp> {
+            }),
+            buildButton('user - add denied GDPR Consent State', () async {
+              var user = await mpInstance?.getCurrentUser();
+              var gdprConsent = Consent(false);
+              gdprConsent.document = 'document test';
+              gdprConsent.hardwareId = 'hardwareID';
+              gdprConsent.location = 'loction test';
+              gdprConsent.timestamp = DateTime.now().millisecondsSinceEpoch;
```
```suggestion
              var gdprConsent = (Consent(false))
                ..document = 'document test'
                ..hardwareId = 'hardwareID'
                ..location = 'loction test'
                ..timestamp = DateTime.now().millisecondsSinceEpoch;
```

rmi22186 replied:
actually i just added this to my PR. So ignore this

rmi22186 commented on example/lib/main.dart:
```diff
@@ -382,6 +383,84 @@ class _MyAppState extends State<MyApp> {
+            }),
+            buildButton('user - add denied CCPA Consent State', () async {
+              var user = await mpInstance?.getCurrentUser();
+              var ccpaConsent = Consent(false);
+              ccpaConsent.document = 'document test';
+              ccpaConsent.hardwareId = 'hardwareID';
+              ccpaConsent.location = 'loction test';
+              ccpaConsent.timestamp = DateTime.now().millisecondsSinceEpoch;
```
```suggestion
              var ccpaConsent = (Consent(false))
                ..document = 'document test'
                ..hardwareId = 'hardwareID'
                ..location = 'loction test'
                ..timestamp = DateTime.now().millisecondsSinceEpoch;
```

rmi22186 replied:
actually i just added this to my PR. So ignore this

rmi22186 reviewed (changes requested):
Minor changes from my side, Will found a [more terse](https://github.com/mParticle/mparticle-flutter-sdk/pull/7#issue-983455787) way to create classes, so I added 2 code suggestions.  Otherwise lgtm!

rmi22186 commented on ios/Classes/SwiftMparticleFlutterSdkPlugin.swift:
```diff
@@ -234,6 +234,114 @@ public class SwiftMparticleFlutterSdkPlugin: NSObject, FlutterPlugin {
+
+               let consentState = MPConsentState()
+               consentState.addGDPRConsentState(consentGDPR, purpose: purpose)
+               if let ccpa = user.consentState()?.ccpaConsentState() {
+                   consentState.setCCPA(ccpa)
+               }
```
actually i was looking at the code again and it looks like you swap CCPA here for GDPR

rmi22186 commented on ios/Classes/SwiftMparticleFlutterSdkPlugin.swift:
```diff
@@ -234,6 +234,114 @@ public class SwiftMparticleFlutterSdkPlugin: NSObject, FlutterPlugin {
+               
+               let consentState = MPConsentState()
+               consentState.setCCPA(consentCCPA)
+               if let gdpr = user.consentState()?.gdprConsentState() {
+                   consentState.setGDPR(gdpr)
+               }
```
GPR and CCPA are swapped here also

BrandonStalnaker replied:
same as the other comment but in reverse.

rmi22186 reviewed (changes requested):
i resolved a few conversations because i added them to my PR [here](https://github.com/mParticle/mparticle-flutter-sdk/pulls#:~:text=Sort-,feat%3A%20Implement%20consent%20mapping%20for%20web,-%2318%20opened%202)
